cordova-plugin-outline/android/java/org/outline/OutlinePlugin.java

```diff
@@ -185,97 +198,92 @@ public boolean execute(String action, JSONArray args, CallbackContext callbackCo
-                    // onActivityResult
-                    startRequestTunnelId = tunnelId;
-                    startRequestConfig = args.getJSONObject(1);
-                    prepareAndStartVpn();
-                  } else if (Action.STOP.is(action)) {
-                    stopVpnTunnel(tunnelId);
-                  } else if (Action.IS_REACHABLE.is(action)) {
-                    boolean isReachable =
-                        ShadowsocksConnectivity.isServerReachable(
-                            args.getString(1), args.getInt(2));
-                    PluginResult result = new PluginResult(PluginResult.Status.OK, isReachable);
-                    sendPluginResult(tunnelId, action, result, false);
-                  } else if (Action.IS_RUNNING.is(action)) {
-                    PluginResult result =
-                        new PluginResult(PluginResult.Status.OK, isTunnelActive(tunnelId));
-                    sendPluginResult(tunnelId, action, result, false);
-
-                    // Static actions
-                  } else if (Action.INIT_ERROR_REPORTING.is(action)) {
-                    final String apiKey = args.getString(0);
-                    OutlineLogger.initializeErrorReporting(getBaseContext(), apiKey);
-                    callback.success();
-                  } else if (Action.REPORT_EVENTS.is(action)) {
-                    final String uuid = args.getString(0);
-                    OutlineLogger.sendLogs(uuid);
-                    callback.success();
-                  } else {
-                    LOG.severe(
-                        String.format(Locale.ROOT, "Unexpected asynchronous action %s", action));
-                    callback.error(ErrorCode.UNEXPECTED.value);
-                  }
-                } catch (Exception e) {
-                  LOG.log(Level.SEVERE, "Unexpected error while executing action.", e);
-                  if (isTunnelInstanceAction(action)) {
-                    PluginResult pluginResult =
-                        new PluginResult(PluginResult.Status.ERROR, ErrorCode.UNEXPECTED.value);
-                    sendPluginResult(tunnelId, action, pluginResult, false);
-                  } else {
-                    callback.error(ErrorCode.UNEXPECTED.value);
-                  }
-                }
-              }
-            });
-  }
-
-  private void prepareAndStartVpn() {
-    if (prepareVpnService()) {
-      startVpnTunnel();
-    }
+  private void executeAsync(
+      final String action, final JSONArray args, final CallbackContext callback) {
+    cordova.getThreadPool().execute(() -> {
+      try {
+        // Tunnel instance actions: tunnel ID is always the first argument.
+        if (Action.START.is(action)) {
+          final String tunnelId = args.getString(0);
+          final JSONObject config = args.getJSONObject(1);
+          int errorCode = startVpnTunnel(tunnelId, config);
+          sendErrorCode(callback, errorCode);
+          startVpnRequest = null;
```

Comment:
I think it would be clearer for this to happen in `onActivityResult` after the call to `executeAsync`.

Reply:
Done